pyop2/compilation.py

```diff
@@ -112,26 +138,38 @@ def get_so(self, src, extension):
         # atomically (avoiding races).
         tmpname = os.path.join(cachedir, "%s_p%d.so.tmp" % (basename, pid))
 
+        if MPI.version >= 3:
+            # Assume that the library supports flashy MPI3 features
+            newcomm = self.comm.Split_type(MPI.COMM_TYPE_SHARED)
```

Comment:
Unfortunately, mpi4py doesn't garbage collect much (in particular communicators).  So this leaks comm handles (one per compile call).  So you either need to call newcomm.Free() later, or else (if we don't want to do all the determination every time we compile things) manage the split communicators yourself using MPI attributes.

PyOP2 already has code for managing "internal" communicators (when the user passes in a communicator, we check to see if we've already dupped it, if not, we dup it and stash the dupped communicator using attributes.  See that code in `pyop2/mpi.py`.  It should be readily adaptable for your usecase (you just need to change the COMM_DUP call to COMM_SPLIT).